src/Internal/Types/Interval.purs

```diff
@@ -333,53 +435,59 @@ lowerBound a = LowerBound (Finite a) true
 upperBound :: forall (a :: Type). a -> UpperBound a
 upperBound a = UpperBound (Finite a) true
 
--- | `interval a b` includes all values that are greater than or equal to `a`
--- | and smaller than or equal to `b`. Therefore it includes `a` and `b`.
-interval :: forall (a :: Type). a -> a -> Interval a
-interval s s' = mkInterval (lowerBound s) (upperBound s')
+-- | `mkFiniteInterval a b` construct the interval [a,b+1)
+mkFiniteInterval :: forall (a :: Type). Ord a => a -> a -> Interval a
+mkFiniteInterval x y = if x <= y then FiniteInterval x y else EmptyInterval
 
-singleton :: forall (a :: Type). a -> Interval a
-singleton s = interval s s
+-- | `singleton a` is an `Interval` that contains just `a`,
+-- | represented as [a,a+1).
+singleton :: forall (a :: Type). Ord a => Semiring a => a -> Interval a
+singleton s = mkFiniteInterval s s
 
 -- | `from a` is an `Interval` that includes all values that are
--- | greater than or equal to `a`.
-from :: forall (a :: Type). a -> Interval a
-from s = mkInterval (lowerBound s) (UpperBound PosInf true)
+-- | greater than or equal to `a`, represented as [a,infinity].
+from :: forall (a :: Type). Ord a => a -> Interval a
```

Comment:
To form an Interval in a set you need at least an order. In this case, we don't need it for computations but is morally wrong to allow to form of an Interval without an Order.

Reply:
In fact, for all the functions that perform some arithmetic (`Semiring` and `Ring`), we should ask for consistency between the order and operations ( x<x+1 is true?).  But encoding this consistency it's not worth the effort for the limited set of cases we need.

Reply:
I suppose. I don't think that argument really works for `always` and `never` though - they're the free (in the same sense Maybe is the free monoid of semigroups). You don't need ord for them because Interval implements them for us

Also: 
```purescript
data NoOrd = NoOrd

iLied :: Interval NoOrd
iLied = from unit <#> const NoOrd
```

I don't mind if you keep it, it does help typing fail faster.

Reply:
Yes, When I introduced the `Generic` instance and tried to hide the constructors, the compiler began to warn about how it could be used to construct intervals. 

In the case of `Generic` we cannot remove it without having problems with other types. But in the case of `Functor`, I introduce it for convenience, we can just use a `mapInterval` function that won't be exported.